### PR TITLE
fix: disable persistent and in-memory cache when loading resources in oss

### PR DESCRIPTION
**Pull request by deepak1556**
Refs https://github.com/microsoft/vscode/issues/148541

### CODE REVIEW

File: src/vs/platform/protocol/electron-main/protocolMainService.ts [hunk 1]
```diff
@@ -104,6 +104,12 @@ export class ProtocolMainService extends Disposable implements IProtocolMainServ
 			}
 		}
 
+		// Disable in-memory cache when running out of sources
+		if (process.env['VSCODE_DEV']) {
```
Comment (bpasero) [inline src/vs/platform/protocol/electron-main/protocolMainService.ts:108]:
```suggestion
		if (!this.environmentService.isBuilt) {
```

File: src/vs/platform/protocol/electron-main/protocolMainService.ts [hunk 2]
```diff
@@ -104,6 +104,12 @@ export class ProtocolMainService extends Disposable implements IProtocolMainServ
 			}
 		}
 
+		// Disable in-memory cache when running out of sources
```
Comment (bpasero) [inline src/vs/platform/protocol/electron-main/protocolMainService.ts:107]: Please reference the issue as to why we do this

File: src/main.ts [hunk 3]
```diff
@@ -304,6 +304,11 @@ function configureCommandlineSwitchesSync(cliArgs: NativeParsedArgs) {
 		app.commandLine.appendSwitch('js-flags', jsFlags);
 	}
 
+	// Disable http persistent disk cache when running out of sources
```
Comment (bpasero) [inline src/main.ts:307]: Please reference the issue as to why we do this

**Review by bpasero — CHANGES_REQUESTED** [review]

**Comment by bpasero** [comment]
I am actually wondering if this makes the development experience worse for engineers on slower machines. So curious if at all its possible to limit this to CSS files only? Why do we need to do changes in 2 files, wouldn't the one in protocol service suffice, where we could do it for CSS Only?

File: src/vs/platform/protocol/electron-main/protocolMainService.ts [hunk 4]
```diff
@@ -104,6 +104,12 @@ export class ProtocolMainService extends Disposable implements IProtocolMainServ
...
+		if (process.env['VSCODE_DEV']) {
+			headers = headers || {};
+			headers['Cache-Control'] = 'no-cache';
```
Comment (Tyriar) [inline src/vs/platform/protocol/electron-main/protocolMainService.ts:110]: `COI.CoopAndCoep` uses `Object.freeze`, is this going to throw a runtime exception?

**Review by Tyriar — CHANGES_REQUESTED** [review]

**Comment by deepak1556** [comment]
There are two types of caches one for persistent and other for in-memory. All resources will first be put into the in-memory cache and after a certain threshold they will get persisted to the disk cache to be read from. There is no finer control for which resource to not use the persistent cache with the current API exposed from the runtime, for the in the in-memory cache since I control it based on header I can isolate it to the resource types we want.

There wouldn't be a performance cost with disabling the persistent cache for oss since we load each resource from disk even without the cache. However, there will be cost when not using the in-memory cache. The issue refers to css files but this caching strategy applies to all resources.

Maybe as a first pass I target the fix per below,

1) Avoid in-memory cache for css files
2) Disable persistent cache globally

**Comment by Tyriar** [comment]
@bpasero I've been sending trace logs and investigating with @deepak1556 the problem I've been seeing and this does seem to fix my JS files consistently not loading problem, so it should significantly improve my situation by allowing me to reload again and not accidentally test without changes which gets very frustrating.

**Comment by deepak1556** [comment]
Given @Tyriar experience, what if we provide a `argv.json` setting to control disabling the in-memory cache for provided resource types, for ex: `disable-http-memory-cache: "js,css,img,html"` ?

**Comment by bpasero** [comment]
Curious: what if we were to file watch changes in OSS and then update the headers accordingly. Is it at all possible to fix this issue like a web server would, by invalidating a cache for a resource?

I am bringing this up because maybe we can fix this nicely enough on a per-file basis even if possible?

**Comment by deepak1556** [comment]
Yeah what you are thinking should be done by the `last-modified` header, the backend in the runtime which performs file loading already sets this https://source.chromium.org/chromium/chromium/src/+/main:content/browser/loader/file_url_loader_factory.cc;l=728-731 based on the file modified time.

Based on the runtime trace,

```
Title | ResourceFetcher::DetermineRevalidationPolicy
Category | blink

policy | "use"
reason | "Use the existing resource because there is no reason not to do so."
```

we are not getting [RevalidationPolicy::kRevalidate](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/loader/fetch/resource_fetcher.h;l=486-492). I need to check why that is the case to come up with a cleaner solution https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/loader/fetch/resource_fetcher.cc;l=2001-2190.

**Comment by deepak1556** [comment]
This smells to be the issue https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/loader/fetch/resource.cc;l=437-440 which avoids calculating cache freshness based on the `last-modified` header https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/loader/fetch/resource_fetcher.cc;l=2148-2186

**Comment by deepak1556** [comment]
Yeah that was it, I am able to see proper cache validation now. I will follow-up with a fix in the runtime.

**Pull request closed without merging**